File: examples/few_body/three_body.py
import numpy as np
import matplotlib.pyplot as plt

class State:
    """ a container to hold a single star's position """

    def __init__(self, x, y, u, v):

        self.x = x
        self.y = y
        self.u = u
        self.v = v

    def __add__(self, other):
        return State(self.x + other.x, self.y + other.y,
                     self.u + other.u, self.v + other.v)

    def __sub__(self, other):
        return State(self.x - other.x, self.y - other.y,
                     self.u - other.u, self.v - other.v)

    def __mul__(self, other):
        return State(other * self.x, other * self.y,
                     other * self.u, other * self.v)

    def __rmul__(self, other):
        return self.__mul__(other)

    def __str__(self):
        return f"{self.x:10.6f} {self.y:10.6f} {self.u:10.6f} {self.v:10.6f}"


class ThreeBody(object):
    """ model the gravitational interaction of three stars """

    def __init__(self, M0, pos0, M1, pos1, M2, pos2, SMALL=1.e-30):
        """ pos is a tuple of x-y coords of the stars """

        self.G = 1  # dimensionless units

        star0 = State(pos0[0], pos0[1], 0.0, 0.0)
        star1 = State(pos1[0], pos1[1], 0.0, 0.0)
        star2 = State(pos2[0], pos2[1], 0.0, 0.0)

        # this history will be stored in a list of tuples, such that
        # stars[n][istar] is star istar and timestep n

        self.stars = [(star0, star1, star2)]

        self.M = np.array([M0, M1, M2])

        self.SMALL = SMALL

        self.time = [0.0]

    def npts(self):
        """ return the number of integration points """
        return len(self.time)

    def center_of_mass(self, n):
        """return the center of mass of the system at time n"""

        Mtot = np.sum(self.M)

        x_cm = 0.0
        y_cm = 0.0

        for mass, star in zip(self.M, self.stars[n]):
            x_cm += mass * star.x
            y_cm += mass * star.y

        x_cm /= Mtot
        y_cm /= Mtot

        return x_cm, y_cm

    def energy(self, n):
        """return the energy of the system at time n """

        # kinetic energy
        KE = [0.5 * self.M[i] * (self.stars[n][i].u**2 +
                                 self.stars[n][i].v**2)
              for i in range(len(self.M))]

        # potential energy -- we need all pairs, but order doesn't matter

        PE = [-self.G * self.M[i] * self.M[j] /
              np.sqrt((self.stars[n][i].x - self.stars[n][j].x)**2 +
                      (self.stars[n][i].y - self.stars[n][j].y)**2)
              for i in range(len(self.M)) for j in range(len(self.M)) if i > j]

        return sum(KE) + sum(PE)

    def rhs(self, star_states):
        """ here, star states is (State, State, ...) """

        ydots = []

        for istar in range(len(self.M)):

            # compute the space derivatives

            dxdt = star_states[istar].u
            dydt = star_states[istar].v

            # compute the velocity derivatives
            dudt = 0.0
            dvdt = 0.0

            for jstar in range(len(self.M)):
                if istar == jstar:
                    continue
                else:
                    dx = star_states[jstar].x - star_states[istar].x
                    dy = star_states[jstar].y - star_states[istar].y

                    r = np.sqrt(dx**2 + dy**2) + self.SMALL

                    dudt += self.G * self.M[jstar] * dx / r**3
                    dvdt += self.G * self.M[jstar] * dy / r**3

            ydots.append(State(dxdt, dydt, dudt, dvdt))

        return ydots


    def single_step(self, state_old, dt):
        """ take a single RK-4 timestep from t to t+dt for the system
        ydot = rhs """

        ydot1 = self.rhs(state_old)

        state_temp = [state_old[istar] + 0.5 * dt * ydot1[istar]
                      for istar in range(len(self.M))]
        ydot2 = self.rhs(state_temp)

        state_temp = [state_old[istar] + 0.5 * dt * ydot2[istar]
                      for istar in range(len(self.M))]
        ydot3 = self.rhs(state_temp)

        state_temp = [state_old[istar] + dt * ydot3[istar]
                      for istar in range(len(self.M))]
        ydot4 = self.rhs(state_temp)

        state_new = [state_old[istar] + (dt/6.0) * (ydot1[istar] + 2.0*ydot2[istar] +
                                                    2.0*ydot3[istar] + ydot4[istar])
                     for istar in range(len(self.M))]

        return state_new

    def integrate(self, dt, err, tmax):
        """ integrate the equations of motion using 4th order R-K
            method with an adaptive stepsize, to try to achieve the
            relative error err.  dt here is the initial timestep

            if err < 0, then we don't do adaptive stepping, but rather
            we always walk at the input dt
            """

        # safety parameters
        S1 = 0.9
        S2 = 4.0

        # initial conditions
        t = 0.0

        # start with the old timestep
        dt_new = dt

        n_reset = 0

        while t < tmax:

            state_old = self.stars[-1]

            # adaptive stepping iteration loop -- keep trying to take
            # a step until we achieve our desired error
            rel_error = 1.e10

            n_try = 0
            while rel_error > err:
                dt = dt_new
                if t + dt > tmax:
                    dt = tmax-t

                # take 2 half steps
                state_tmp = self.single_step(state_old, 0.5*dt)
                state_new = self.single_step(state_tmp, 0.5*dt)

                # now take just a single step to cover dt
                state_single = self.single_step(state_old, dt)

                # state_new should be more accurate than state_single,
                # since it used smaller steps.

                # estimate the relative error now
                pos_err = max([max(abs((state_new[istar].x - state_single[istar].x) / state_single[istar].x),
                                   abs((state_new[istar].y - state_single[istar].y) / state_single[istar].y))
                               for istar in range(len(self.M))])

                vel_err = max([max((abs(state_new[istar].u - state_single[istar].u) / state_single[istar].u),
                                   (abs(state_new[istar].v - state_single[istar].v) / state_single[istar].v))
                               for istar in range(len(self.M))])

                rel_error = max(pos_err, vel_err)

                # adaptive timestep algorithm from Garcia (Eqs. 3.30
                # and 3.31)
                dt_est = dt * abs(err / rel_error)**0.2
                dt_new = min(max(S1*dt_est, dt/S2), S2*dt)

                n_try += 1

            if n_try > 1:
                # n_try = 1 if we took only a single try at the step
                n_reset += (n_try-1)

            # successful step
            t += dt

            # set for the next step
            self.time.append(t)
            self.stars.append(state_new)

            print(t, end="")
            for s in state_new:
                print(s, end="")
            print("\n")


if __name__ == "__main__":

    tb = ThreeBody(150., (3, 1),
                   200., (-1, -2),
                   250., (-1, 1))

    tb.integrate(0.05, 1.e-8, 2.0)

    print(f"number of steps: {tb.npts()}")

    fig = plt.figure()
    ax = fig.add_subplot()

    for istar in range(len(tb.M)):
        x = [star[istar].x for star in tb.stars]
        y = [star[istar].y for star in tb.stars]

        ax.plot(x, y)
        ax.scatter([x[0]], [y[0]], marker="x")

    fig.savefig("three_body.png")

    plt.close(fig)

    fig = plt.figure()
    ax = fig.add_subplot()

    t = tb.time
    x_cm = []
    y_cm = []
    for n in range(tb.npts()):
        _xcm, _ycm = tb.center_of_mass(n)
        x_cm.append(_xcm)
        y_cm.append(_ycm)

    ax.plot(t, x_cm, label="x center of mass")
    ax.plot(t, y_cm, label="y center of mass")

    fig.savefig("center_of_mass.png")

    plt.close(fig)

    # timestep plot

    t = np.asarray(t)
    dt = t[1:] - t[:-1]
    ts = 0.5 * (t[1:] + t[:-1])

    fig = plt.figure()
    ax = fig.add_subplot()

    ax.plot(ts, dt)
    ax.set_yscale("log")

    ax.set_xlabel("t")
    ax.set_ylabel("dt")

    fig.savefig("three_body_dts.png")

    plt.close(fig)

    # energy

    fig = plt.figure()
    ax = fig.add_subplot()

    E = []
    for n in range(tb.npts()):
        E.append(tb.energy(n))

    ax.plot(t, np.abs(E))

    ax.set_yscale("log")

    ax.set_xlabel("t")
    ax.set_ylabel("|E|")

    fig.tight_layout()
    fig.savefig("three_body_energy.png")

File: examples/few_body/three_body.cpp
#include <iostream>
#include <vector>
#include <iomanip>
#include <cmath>
#include <numeric>
#include <limits>
#include <fstream>

const double G = 1.0;
const int N = 3;

struct State {
    // a container to hold the star positions
    double x{};
    double y{};
    double u{};
    double v{};

    State(double x_, double y_, double u_, double v_) :
        x(x_), y(y_), u(u_), v(v_)
    {}

    State() {}

};

std::ostream& operator<< (std::ostream& os, const State& s) {
    os.precision(6);

    os << std::setw(14) << s.x
       << std::setw(14) << s.y
       << std::setw(14) << s.u
       << std::setw(14) << s.v;

    return os;
}

class ThreeBody {
    /// model the gravitational interaction of three stars

private:

    double SMALL;
    std::vector<double> masses;
    std::vector<double> time;
    std::vector<std::vector<State>> stars;
    int n_reset{0};

public:

    ThreeBody (const double M0, const std::vector<double>& pos0,
               const double M1, const std::vector<double>& pos1,
               const double M2, const std::vector<double>& pos2,
               const double _SMALL=1.e-30)
               : SMALL{_SMALL},
                 masses{M0, M1, M2}
    {

        // we will organize the data as stars[n][istar].{t,x,y,u,v}, where i
        // istar the index of the star (0, 1, 2) and n is the timestep index

        std::vector<State> current;

        current.emplace_back(pos0[0], pos0[1], 0.0, 0.0);
        current.emplace_back(pos1[0], pos1[1], 0.0, 0.0);
        current.emplace_back(pos2[0], pos2[1], 0.0, 0.0);

        stars.push_back(current);

        time.push_back(0.0);

    }

    int npts() {
        // return the number of integration points
        return stars.size();
    }

    double get_time(const int n) {
        // return the physical time for step n
        return time[n];
    }

    std::vector<State>& get_state(const int n) {
        // return a reference to the state at time index n
        return stars[n];
    }

    std::pair<double, double> center_of_mass(const int n) {
        // return the center of mass of the system at time n

        auto Mtot = std::accumulate(masses.cbegin(), masses.cend(), 0.0);

        const auto& state = stars[n];

        double xcm{};
        double ycm{};

        for (int istar = 0; istar < N; ++istar) {
            xcm += masses[istar] * state[istar].x;
            ycm += masses[istar] * state[istar].y;
        }
        xcm /= Mtot;
        ycm /= Mtot;

        return {xcm, ycm};
    }

    double energy(const int n) {
        // return the energy of the system for timestep n

        const auto& state = stars[n];

        // kinetic energy

        double KE = 0.0;
        for (int istar = 0; istar < N; ++istar) {
            KE += 0.5 * masses[istar] *
                (std::pow(state[istar].u, 2) +
                 std::pow(state[istar].v, 2));
        }

        // potential energy -- we need all pairs, but order doesn't matter

        double PE = 0.0;
        for (int istar = 0; istar < N; ++istar) {
            for (int jstar = 0; jstar < istar; ++jstar) {
                PE += -G * masses[istar] * masses[jstar] /
                    std::sqrt(std::pow(state[istar].x - state[jstar].x, 2) +
                              std::pow(state[istar].y - state[jstar].y, 2));
            }
        }

        return KE + PE;
    }

    std::vector<State> rhs(const std::vector<State>& star_states) {
        // compute the ydot terms

        std::vector<State> ydot;

        for (int istar = 0; istar < N; ++istar) {
            State star_ydot;

            star_ydot.x = star_states[istar].u;
            star_ydot.y = star_states[istar].v;

            // for the gravitational acceleration, we need the force of this
            // star with all others

            double a_x{};
            double a_y{};

            for (int jstar = 0; jstar < N; ++jstar) {

                // no self-interaction

                if (istar == jstar) {
                    continue;
                }

                double dx = star_states[jstar].x - star_states[istar].x;
                double dy = star_states[jstar].y - star_states[istar].y;

                double r = std::sqrt(dx * dx + dy * dy) + SMALL;

                a_x += G * masses[jstar] * dx / std::pow(r, 3);
                a_y += G * masses[jstar] * dy / std::pow(r, 3);

            }

            star_ydot.u = a_x;
            star_ydot.v = a_y;

            ydot.push_back(star_ydot);

        }

        return ydot;

    }

    std::vector<State> update_state(const std::vector<State>& state_old,
                                    const std::vector<State>& ydot,
                                    const double dt) {

        std::vector<State> state_new;

        for (int istar = 0; istar < N; ++istar) {
            State s;
            s.x = state_old[istar].x + dt * ydot[istar].x;
            s.y = state_old[istar].y + dt * ydot[istar].y;
            s.u = state_old[istar].u + dt * ydot[istar].u;
            s.v = state_old[istar].v + dt * ydot[istar].v;

            state_new.push_back(s);
        }

        return state_new;
    }

    std::vector<State> single_step(const std::vector<State>& state_old, const double dt) {
        /// take a single RK-4 timestep through dt

        auto ydot1 = rhs(state_old);

        std::vector<State> state_temp;

        state_temp = update_state(state_old, ydot1, 0.5*dt);
        auto ydot2 = rhs(state_temp);

        state_temp = update_state(state_old, ydot2, 0.5*dt);
        auto ydot3 = rhs(state_temp);

        state_temp = update_state(state_old, ydot3, dt);
        auto ydot4 = rhs(state_temp);

        // final solution

        std::vector<State> state_new;

        for (int is = 0; is < N; ++is) {
            State s;

            s.x = state_old[is].x + (dt / 6.0) *
                (ydot1[is].x + 2.0 * ydot2[is].x + 2.0 * ydot3[is].x + ydot4[is].x);
            s.y = state_old[is].y + (dt / 6.0) *
                (ydot1[is].y + 2.0 * ydot2[is].y + 2.0 * ydot3[is].y + ydot4[is].y);
            s.u = state_old[is].u + (dt / 6.0) *
                (ydot1[is].u + 2.0 * ydot2[is].u + 2.0 * ydot3[is].u + ydot4[is].u);
            s.v = state_old[is].v + (dt / 6.0) *
                (ydot1[is].v + 2.0 * ydot2[is].v + 2.0 * ydot3[is].v + ydot4[is].v);

            state_new.push_back(s);
        }

        return state_new;
    }


    void integrate(const double dt_in, const double err, const double tmax) {
        // integrate the equations of motion using 4th order R-K method with an
        // adaptive stepsize, to try to achieve the relative error err.  dt here
        // is the initial timestep

        // if err < 0, then we don't do adaptive stepping, but rather
        // we always walk at the input dt

        // start with the old timestep
        double dt_new{dt_in};
        double dt{dt_in};

        double t{0.0};

        while (t < tmax) {

            const auto &state_old = stars.back();

            // adaptive stepping iteration loop -- keep trying to take
            // a step until we achieve our desired error
            double rel_error = std::numeric_limits<double>::max();

            int n_try{0};

            std::vector<State> state_new;

            while (rel_error > err) {
                dt = std::min(dt_new, tmax - t);

                // take 2 half steps

                auto state_temp = single_step(state_old, 0.5*dt);
                state_new = single_step(state_temp, 0.5*dt);

                // now take just a single step to cover dt

                auto state_single = single_step(state_old, dt);

                // state_new should be more accurate than state_single since it
                // used smaller steps

                // estimate the relative error now

                double pos_err = std::numeric_limits<double>::lowest();
                double vel_err = std::numeric_limits<double>::lowest();

                for (int is = 0; is < N; ++is) {
                    pos_err = std::max(pos_err,
                                       std::abs((state_new[is].x - state_single[is].x) /
                                                state_single[is].x));
                    pos_err = std::max(pos_err,
                                       std::abs((state_new[is].y - state_single[is].y) /
                                                state_single[is].y));

                    vel_err = std::max(vel_err,
                                       std::abs((state_new[is].u - state_single[is].u) /
                                                state_single[is].u));
                    vel_err = std::max(vel_err,
                                       std::abs((state_new[is].v - state_single[is].v) /
                                                state_single[is].v));
                }

                rel_error = std::max(pos_err, vel_err);

                // adaptive timestep algorithm from Garcia (Eqs. 3.30 and 3.31)

                double dt_est = dt * std::pow(std::abs(err/rel_error), 0.2);
                dt_new = std::min(std::max(0.9 * dt_est, 0.25 * dt), 4.0 * dt);

                n_try++;
            }

            if (n_try > 1) {
                // n_try = 1 if we took only a single try at the step
                n_reset += (n_try - 1);
            }

            // successful step
            t += dt;

            // store the new solution -- this will automatically be used as the
            // "old" state in the next step

            time.push_back(t);
            stars.push_back(state_new);

        }

    }
};


int main() {

    std::vector<double> pos0{3.0, 1.0};
    std::vector<double> pos1{-1.0, -2.0};
    std::vector<double> pos2{-1.0, 1.0};

    double M0 = 150.0;
    double M1 = 200.0;
    double M2 = 250.0;

    ThreeBody tb(M0, pos0, M1, pos1, M2, pos2);
    tb.integrate(0.05, 1.e-8, 2.0);


    std::cout << "number of points = " << tb.npts() << std::endl;

    std::ofstream of("three_body.dat");

    for (int n = 0; n < tb.npts(); ++n) {
        const auto& state = tb.get_state(n);

        of << std::setw(14) << tb.get_time(n);

        // first the positions and velocities

        for (const auto& star : state) {
            of << star;
        }

        // now the energy

        of.precision(10);

        of << std::setw(18) << tb.energy(n);

        // and the center of mass

        auto [xcm, ycm] = tb.center_of_mass(n);

        of << std::setw(18) << xcm << std::setw(18) << ycm;

        of << std::endl;
    }

}
